### PR TITLE
feat: Add form date fields

### PR DESCRIPTION
**Pull request by paulgain**
WIP: Can you guys cast your eyes over this PR to ensure I'm on the right path. I've got lots more tests to write as well as adding more attributes like value, validation etc etc.

html

### CODE REVIEW

File: src/forms/elements/FieldDate.jsx [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+
+import FieldInput from './FieldInput'
+
+const FieldDate = ({ name, legend }) => {
+  return (
+    <fieldset>
+      {legend && <legend>{legend}</legend>}
+      <FieldInput name={`${name}-day`} type="number" label="Day" pattern="[0-9]*" />
```
Comment (rafenden) [inline src/forms/elements/FieldDate.jsx]: I suggest to use a normal `<input />` as each field should return only one value. i.e. 
**Expected:**
```
{ myDateField: '2019-08-19' }
```
**Got:**
```
{
  myDateField-day: '19',
  myDateField-month: '08',
  myDateField-year: '2019',
}
```

File: src/forms/elements/FieldDate.jsx [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+
+import FieldInput from './FieldInput'
+
+const FieldDate = ({ name, legend }) => {
```
Comment (rafenden) [inline src/forms/elements/FieldDate.jsx]: I suggest to use `label` instead of `legend` to be consistent with `FieldInput`.

**Comment by paulgain** [comment]
@rafenden I've only just read your review. I posted this last Friday, this morning there was no response so I continued on the path I was on. In future, I suggest you give feedback earlier as this is not helpful once it's been developed.

File: src/forms/journeys/MultistepExampleForm.jsx [hunk 3]
```diff
@@ -38,6 +39,13 @@ function MultistepExampleForm() {
           <Step name="first">
             <FieldInput type="number" name="age" label="Age" validate={value => (!value ? 'Age is required' : null)} />
             <FieldInput type="number" name="height" label="Height" validate={value => (!value ? 'Height is required' : null)} />
+            <FieldDate
+              name="dob"
+              legend="Date of birth"
+              validateDay={value => (!value ? 'Day is required' : null)}
```
Comment (rafenden) [inline src/forms/journeys/MultistepExampleForm.jsx]: I suggest having a single validation callback and then pass the complete date as value, same as `FieldInput`.

File: src/forms/elements/FieldDate.jsx (hunk 2 above)
Comment (paulgain) [reply src/forms/elements/FieldDate.jsx]: 👍

File: src/forms/journeys/MultistepExampleForm.jsx (hunk 3 above)
Comment (paulgain) [reply src/forms/journeys/MultistepExampleForm.jsx]: 👍

File: src/forms/elements/FieldDate.jsx (hunk 1 above)
Comment (paulgain) [reply src/forms/elements/FieldDate.jsx]: 👍

**Comment by paulgain** [comment]
Currently form validation is performed when the user clicks next. I'm  unsure if this is the best approach considering we're using React - we have everything we need within the browser. The current approach seems to mimic our existing implementation of Node/Express/Nunjucks where
the validation is done server side when the user clicks next - I don't know if this is intentional.

 We could improve client side validation by validating a field once it has lost focus rather than waiting for the user to complete the entire form and then informing them where they've gone wrong. I don't know how much effort it would be to make those changes, we're at an early stage so it should be minimal.

I've applied validation to each of the day fields (day, month and year). The validation is applied `onBlur` (element has lost focus) therefore the user receives immediate feedback as just mentioned.

**Comment by web-bert** [comment]
GDS to not recommend immediate validation:

> Only display an error when someone tries to move to the next part of the service. Do not show error messages:
> * when users select or tab to a field
> * as they are typing
> * when they move away from a field

https://design-system.service.gov.uk/components/error-message/

**Comment by paulgain** [comment]
> GDS do not recommend immediate validation:
> 
> > Only display an error when someone tries to move to the next part of the service. Do not show error messages:
> > 
> > * when users select or tab to a field
> > * as they are typing
> > * when they move away from a field
> 
> https://design-system.service.gov.uk/components/error-message/

👍 

GDS also

**Pull request closed without merging**